Init pmon from resource manager and read resourceEvents from kafka

GeorgeJahad opened the pull request:
# Resolves
https://jira.rax.io/browse/SALUS-135
https://jira.rax.io/browse/SALUS-117

# What
Replaces the expected keys read/watched in etcd, with the resources recieved from the resourceManager, and the kafka resourceEvent topic


## How to test
Run the tests
(Note that the tests use the embedded kafka broker in spring.)

# TODO
need to get the correct resource manager url, and figure out the hasEnvoy parameter on the  getResource() rest api call.

itzg commented on src/main/resources/application.yml:
```diff
@@ -17,3 +24,16 @@ spring:
+  exportPeriod:
+    1
+  resourceManagerUrl:
+    http://localhost/getResources
```
Is this just a stub/default value or an actual location that is served by something?

itzg commented on src/test/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessorTest.java:
```diff
@@ -56,11 +85,10 @@ MeterRegistry getMeterRegistry() {
 
-    @MockBean
-    MetricRouter metricRouter;
+    private MetricRouter metricRouter;
```
You could annotate that with a regular Mockito `@Mock` and avoid having to call `Mockito.mock(MetricRouter.class)` in each unit test.

GeorgeJahad replied:
@itzg But then the same mock gets reused in each test, causing the test to fail, because i need a fresh one in each test.  Am I misunderstanding something?

GeorgeJahad commented on src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/ResourceListener.java:
```diff
@@ -0,0 +1,96 @@
+        }
+    }
+
+    BiConsumer<PartitionSlice, ConsumerRecord<String, ResourceEvent>> updateSlice = (slice, record) -> {
```
@itzg I created the biConsumer so the test could wrap the biConsumer in another one that generates a semaphore.  (I don't like using sleep's in my test code.)  I can't do that with a regular method, (I don't think.)  Have I designed the test incorrectly?  How would you do it?

itzg replied:
Agreed, sleep's are very good to avoid. I would use Mockito's spy (and delegating) features by spying on the ResourceListener instance

https://static.javadoc.io/org.mockito/mockito-core/2.23.4/org/mockito/Mockito.html#27

jjbuchan commented on src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/ResourceListener.java:
```diff
@@ -0,0 +1,96 @@
+            String key = record.key();
+            ResourceEvent resourceEvent = record.value();
+            ResourceInfo rinfo = PresenceMonitorProcessor.convert(resourceEvent.getResource());
+            if (!resourceEvent.getOperation().equalsIgnoreCase("delete")) {
```
https://github.com/racker/salus-telemetry-model/pull/22/files#diff-d8910738c5ff35ce0a74bc6d9d5ac8f1

GeorgeJahad commented:
@itzg I believe I addressed all your issues with the following exceptions:

I couldn't get spring.application.name to work and found no reference to it being the default group id in the docs, so I just used this instead:
spring.kafka.consumer.group-id: presence-monitor-${random.value}


Also, I still need to work out the hashing with James.

The diffs are here:
https://github.com/racker/salus-telemetry-presence-monitor/compare/63d035918a3f5d63088bb47018d2229a4184bfc8...4830320980c83b0faafc3be245069f1c1cc43f51

GeorgeJahad commented:
I had to add more changes based on my integration tests with james.

Things seem to be working now.  Basically I had to add the hashing, and restructure the resourcelistener to be a bean, because i stupidly failed to remember the kafka auto config only works with beans.  Anyway if you just want to see those diffs, they are here:
https://github.com/racker/salus-telemetry-presence-monitor/compare/4830320...c47106a

itzg commented:
Aha, I was confusing `spring.application.name` with a Kafka Streams config thing:
> Enabling Kafka Streams means that the application id and bootstrap servers must be set. The former can be configured using spring.kafka.streams.application-id, defaulting to spring.application.name if not set.
from [here](https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-messaging.html#boot-features-kafka-streams)

itzg commented on src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java:
```diff
@@ -0,0 +1,43 @@
+
+
+@Configuration
+public class ResourceListenerBean {
```
Nit: I prefer `@Configuration` classes to end with "Config"

itzg commented on src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Bean
+    public ConcurrentHashMap<String, PartitionSlice> getPartitionTable() {
+        return partitionTable;
```
`@Bean` methods are basically factory methods, so the implicit contract is that you "new" the bean inside the method. If you're wanting to ensure there's only ever one (guessing from the `static` declaration), rest assured that by default Spring bean/components are singletons and Spring will guarantee that your factory method is called only once.

GeorgeJahad replied:
@itzg The problem is I want the PresenceMonitorProcessor class to share the partitionTable with the resourceListener

itzg replied:
They would both get the same, singleton instance. Spring will only call the `@Bean` factory once and then provide the returned instance to all uses of it.

itzg commented on src/main/java/com/rackspace/salus/telemetry/presence_monitor/config/ResourceListenerBean.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @Bean("resourceListener")
+    public ResourceListener getResourceListener() {
```
The convention is actually to name a `@Bean` method in non-getter former, like `resourceListener` in this case. Cool thing in this case is that you you then don't need the name specified in `@Bean`.

GeorgeJahad replied:
When I remove it I get this error:
Unsatisfied dependency expressed through constructor parameter 9; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type 'com.rackspace.salus.telemetry.presence_monitor.services.ResourceListener' available: expected at least 1 bean which qualifies as autowire candidate. Dependency annotations: {@org.springframework.beans.factory.annotation.Qualifier(value=resourceListener)}

itzg replied:
You'll need to make both changes I suggested there to end up with

```
    @Bean
    public ResourceListener resourceListener() {
        return new ResourceListener(partitionTable);
    }
```

itzg replied:
FYI the bean naming behavior of `@Bean` is described deep in the docs here

https://docs.spring.io/spring/docs/5.1.4.RELEASE/spring-framework-reference/core.html#beans-java-bean-annotation

itzg commented on src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java:
```diff
@@ -84,16 +86,17 @@
                              MeterRegistry meterRegistry, KeyHashing hashing,
-                             PresenceMonitorProperties props, RestTemplateBuilder restTemplateBuilder) {
+                             PresenceMonitorProperties props, RestTemplateBuilder restTemplateBuilder,
+                             @Qualifier("resourceListener") ResourceListener resourceListener, ConcurrentHashMap<String, PartitionSlice> partitionTable) {
```
Hmm, why did you need the `@Qualifier`? Was it complaining about ambiguous instances?

GeorgeJahad replied:
yes, the test code also creates a bean so that spring can autoconfig it with embedded kafka broker.

itzg commented on src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/ResourceListener.java:
```diff
@@ -43,11 +47,13 @@ public void resourceListener(ConsumerRecord<String, ResourceEvent> record) {
-                    (record.key().compareTo(slice.getRangeMax()) <= 0)) {
+            ResourceInfo rinfo = PresenceMonitorProcessor.convert(record.value().getResource());
+            String hash = PresenceMonitorProcessor.genExpectedId(rinfo);
+            if ((hash.compareTo(slice.getRangeMin()) >= 0) &&
```
Sorry, I forgot to mention this yesterday, but it would be handy to have a method on `PartitionSlice` like a `containsHash` that will do this range conditional. Then you can simply these statements.

itzg commented on src/main/resources/application.yml:
```diff
@@ -7,11 +7,11 @@ presence-monitor:
       telemetry.events.json
     RESOURCE:
-      telemetry.resources.json
+      resources
```
Actually, I prefer the `telemetry.` prefixing type structure of topic name. I keep forgetting to mention to @jjbuchan about the semi-established topic naming convention.

GeorgeJahad replied:
yes, i just changed it for the integration test with james code.

jjbuchan replied:
locally i've changed to telemetry.resources.json and telemetry.attaches.json

itzg commented on src/test/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessorTest.java:
```diff
@@ -127,11 +126,19 @@ MeterRegistry getMeterRegistry() {
 
     @Before
     public void setUp() throws Exception {
+        partitionTable.forEachKey(1, k -> partitionTable.remove(k));
```
You could also use `@DirtiesContext` on the test class or test methods to indicate that a fresh set of components is needed for each test.

If staying with this approach, which is totally fine, I'd recommend just use the [clear method](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#clear--)

GeorgeJahad commented:
ok, i think I got everything:

diffs here:
https://github.com/racker/salus-telemetry-presence-monitor/compare/c47106a0b294d467f725f828e623f0f2adeada18...george/addResourceManager3

GeorgeJahad commented:
@itzg Thank you for taking the time to teach me all of this stuff.  There are so many subtle spring details that I would have missed if not for you.

jjbuchan commented on src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java:
```diff
@@ -70,6 +112,51 @@ private String getExpectedId(KeyValue kv) {
+    }
+
+    static ResourceInfo convert(Resource resource) {
+        ResourceInfo ri = new ResourceInfo();
```
Doesn't need to change, but I did like the additional features this provided - https://github.com/racker/salus-telemetry-resource-management/pull/1/files#diff-d034cbc23c60da4b9dc850475e9f9768R218

jjbuchan commented on src/main/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessor.java:
```diff
@@ -95,23 +183,24 @@ public void start(String id, String content) {
+
+        log.debug("Found {} expected envoys", resources.size());
+        resources.forEach(resource -> {
+            if (!resource.getPresenceMonitoringEnabled())
```
The endpoint your hitting only returns resources with this enabled, so this check should be irrelevant

GeorgeJahad replied:
will do

jjbuchan commented on src/main/java/com/rackspace/salus/telemetry/presence_monitor/types/KafkaMessageType.java:
```diff
@@ -21,5 +21,6 @@
     METRIC,
-    EVENT
+    EVENT,
+    RESOURCE
```
There is a shared `KafkaMessageType` in the models repo you can use.  Under `com.rackspace.salus.telemetry.messaging`

GeorgeJahad replied:
will do

jjbuchan commented on src/test/java/com/rackspace/salus/telemetry/presence_monitor/services/PresenceMonitorProcessorTest.java:
```diff
@@ -1,49 +1,79 @@
+@EnableAutoConfiguration
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.NONE)
-
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
```
Interesting
